titus-ext/cassandra/src/main/java/com/netflix/titus/ext/cassandra/store/CassandraLoadBalancerStore.java

```diff
@@ -274,9 +322,69 @@ public int getNumLoadBalancersForJob(String jobId) {
         return loadBalancerCount;
     }
 
-    private Pair<JobLoadBalancer, JobLoadBalancer.State> buildLoadBalancerStatePairFromRow(Row row) {
-        return Pair.of(new JobLoadBalancer(row.getString(COLUMN_JOB_ID), row.getString(COLUMN_LOAD_BALANCER)),
-                JobLoadBalancer.State.valueOf(row.getString(COLUMN_STATE)));
+    @Override
+    public Mono<Void> addOrUpdateTargets(Collection<LoadBalancerTargetState> targets) {
+        List<Mono<Void>> insertOperations = targets.stream()
+                .map(this::addOrUpdateTarget)
+                .collect(Collectors.toList());
+
+        int limit = configuration.getLoadBalancerConcurrencyLimit();
+        // prefetch does not matter here because operations don't produce any result (they are Mono<Void>)
+        return Flux.mergeSequentialDelayError(insertOperations, limit, limit)
+                .ignoreElements()
+                .doOnSubscribe(ignored -> {
+                    Map<String, Long> countPerLoadBalancer = targets.stream().collect(Collectors.groupingBy(
+                            target -> target.getLoadBalancerTarget().getLoadBalancerId(),
+                            Collectors.counting()
+                    ));
+                    logger.info("Inserting/updating targets: {}", countPerLoadBalancer);
+                    logger.debug("Inserting/updating {} targets. Details: {}", targets.size(), targets);
+                });
+    }
+
+    private Mono<Void> addOrUpdateTarget(LoadBalancerTargetState target) {
+        Set<ValidationError> violations = entitySanitizer.validate(target);
+        if (!violations.isEmpty()) {
+            if (configuration.isFailOnInconsistentLoadBalancerData()) {
+                return Mono.error(LoadBalancerStoreException.badData(target, violations));
+            }
+            logger.warn("Ignoring bad target record of {} due to validation constraint violations: violations={}", target, violations);
+            return Mono.empty();
+        }
+
+        BoundStatement statement = insertTarget.bind(
```

Comment:
If `addOrUpdateTargets` would retry, this statement would be reused which is probably incorrect. Use `Mono.defer(...)`?

Reply:
According to the documentation:
```This class is not thread-safe. Do not share instances among requests that will execute concurrently (e.g. requests run from separate application threads, but also separate executeAsync calls, even if they're triggered from the same thread).```
If we can guarantee strict non-concurrency, we could leave it like that, but it is probably easier to create a new instance for each attempt.

Reply:
Good catch. I will fix.